### PR TITLE
Make Blind fail earlier for the POPRF

### LINKED ISSUE #307: Unhandled zero scalars and identity elements
This is a list of functions that contain group operations that might produce zero scalars or identity elements. I don't actually know if this is a problem or not in some of the cases, I'm going to leave that to actual cryptographers to figure out.

- [`GenerateProof`](https://cfrg.github.io/draft-irtf-cfrg-voprf/draft-irtf-cfrg-voprf.html#section-2.2.1-2) can return a zero scalar because of `s = (r - c * k) mod p`.

- [`ComputeComposites`](https://cfrg.github.io/draft-irtf-cfrg-voprf/draft-irtf-cfrg-voprf.html#section-2.2.2-4) and [`ComputeCompositesFast`](https://cfrg.github.io/draft-irtf-cfrg-voprf/draft-irtf-cfrg-voprf.html#section-2.2.1-4) use `HashToScalar`, which can return a zero scalar, which is then used in `M = di * Cs[i] + M` and `Z = di * Ds[i] + Z`. Adding `M`, prevents producing an identity point, so it might not be a problem.

- POPRF's [`Finalize`](https://cfrg.github.io/draft-irtf-cfrg-voprf/draft-irtf-cfrg-voprf.html#section-3.3.3-4) uses `HashToScalar`, which can return a zero scalar, which is then used in `T = ScalarBaseMult(m)`, which would produce the identity point. Afterwards, in `U = T + pkS`, adding a non-identity point would prevent `U` from becoming an identity point, so it might not be a problem.

### PR DESCRIPTION
**Pull request by chris-wood**
Note that this check on the client is equivalent to the server, since `(skS + t) == 0` is only true iff `pkS + tG = skS*G + tG = (skS + t)G` is the identity element.

Closes #307.

cc @bytemare, @kevinlewi

### CODE REVIEW

**Comment by chris-wood** [comment]
(Please ignore @bifurcation!)

File: draft-irtf-cfrg-voprf.md [hunk 1]
```diff
@@ -983,11 +983,47 @@ def Finalize(input, blind, evaluatedElement, blindedElement, proof):
...
+~~~
+Input:
+
+  PrivateInput input
+  PublicInput info
+
+Output:
+
+  Scalar blind
+  Element blindedElement
```
Comment (chris-wood) [inline draft-irtf-cfrg-voprf.md:1002]:
```suggestion
  Element blindedElement
```
```suggestion
  Element blindedElement
  Element tweakedKey
```

File: draft-irtf-cfrg-voprf.md [hunk 2]
```diff
@@ -983,11 +983,48 @@ def Finalize(input, blind, evaluatedElement, blindedElement, proof):
...
+  P = GG.HashToGroup(input)
+  if P == GG.Identity():
+    raise InvalidInputError
+
+  blindedElement = blind * P
+
+  return blind, blindedElement, tweakedKey
+~~~
+
+Clients store the outputs `blind` and `U` locally and send `blindedElement` to
```
Comment (chris-wood) [inline draft-irtf-cfrg-voprf.md:1025]:
```suggestion
Clients store the outputs `blind` and `tweakedKey` locally and send `blindedElement` to
```

File: draft-irtf-cfrg-voprf.md [hunk 3]
```diff
@@ -1032,19 +1069,14 @@ Input:
...
 
-Errors: InverseError
-
-def Finalize(input, blind, evaluatedElement, blindedElement, proof, info):
-  context = "Info" || I2OSP(len(info), 2) || info
-  m = GG.HashToScalar(context)
-  T = ScalarBaseMult(m)
-  U = T + pkS
-  if VerifyProof(G, U, evaluatedElement, blindedElement, proof) == false:
+def Finalize(input, blind, evaluatedElement, blindedElement, proof, info, U):
```
Comment (chris-wood) [inline draft-irtf-cfrg-voprf.md:1078]:
```suggestion
def Finalize(input, blind, evaluatedElement, blindedElement, proof, info, tweakedKey):
```

File: draft-irtf-cfrg-voprf.md [hunk 4]
```diff
@@ -983,11 +983,48 @@ def Finalize(input, blind, evaluatedElement, blindedElement, proof):
...
 
-The POPRF protocol begins with the client blinding its input, using the same
-`Blind` function as in {{oprf}}. Clients store the output `blind` locally
-and send `blindedElement` to the server for evaluation. Upon receipt, servers
-process `blindedElement` to compute an evaluated element and DLEQ proof using
-the following `Evaluate` function.
+The POPRF protocol begins with the client blinding its input, using the
+following modified `Blind` function. Note that this function can fail with an
+`InvalidInputError` error for certain private inputs that map to the group
+identity element, as well as certain public inputs that map lead to invalid
```
Comment (armfazh) [inline draft-irtf-cfrg-voprf.md:989]:
```suggestion
identity element, as well as certain public inputs that map to invalid
```

File: draft-irtf-cfrg-voprf.md [hunk 5]
```diff
@@ -983,11 +983,48 @@ def Finalize(input, blind, evaluatedElement, blindedElement, proof):
...
+
+  return blind, blindedElement, tweakedKey
+~~~
+
+Clients store the outputs `blind` and `tweakedKey` locally and send `blindedElement` to
+the server for evaluation. Upon receipt, servers process `blindedElement` to
+compute an evaluated element and DLEQ proof using the following `Evaluate` function.
 
 ~~~
 Finalize
```
Comment (armfazh) [inline draft-irtf-cfrg-voprf.md:1030]:
```suggestion
```
--remove Finalize

File: draft-irtf-cfrg-voprf.md [hunk 6]
```diff
@@ -983,11 +983,48 @@ def Finalize(input, blind, evaluatedElement, blindedElement, proof):
...
+
+Errors: InvalidInputError
+
+def Blind(input, info):
+  context = "Info" || I2OSP(len(info), 2) || info
+  m = GG.HashToScalar(context)
+  T = ScalarBaseMult(m)
+  tweakedKey = T + pkS
+  if tweakedKey == GG.Identity():
+    raise InvalidInputError
```
Comment (armfazh) [inline draft-irtf-cfrg-voprf.md:1013]: once client arrives to this line, it follows 
```
T + pkS     == 0
T + skS * G == 0
skS * G     == -T
skS * G     == - m * G
skS         == -m
```
and the server's private key is found.
is this a concern for the server?

Comment (chris-wood) [reply draft-irtf-cfrg-voprf.md:1013]: No, I don't think so, since it's an exceptional case and there's nothing a server can really do to prevent it.

**Review by armfazh — APPROVED** [review]

**Pull request merged**